File: python/private_key/6_market_order.py
# python 6_market_order.py  — FOK BUY at best ask → poll fill → FOK SELL back
import os, time, requests
from dotenv import load_dotenv
from web3 import Web3
from py_clob_client_v2.client import ClobClient
from py_clob_client_v2.clob_types import (
    OrderArgsV2, PartialCreateOrderOptions, OrderType,
    BalanceAllowanceParams, AssetType,
)

load_dotenv()

PRIVATE_KEY = os.environ["PRIVATE_KEY"]
CLOB_HOST   = os.environ.get("CLOB_BASE_URL", "https://clob-v2.polymarket.com")
RPC_URL     = os.environ.get("POLYGON_RPC_URL", "https://polygon-bor-rpc.publicnode.com")
EVENT_ID    = "73106"
PUSD        = Web3.to_checksum_address("0xC011a7E12a19f7B1f670d46F03B03f3342E82DFB")
ERC20_ABI   = [{"name": "balanceOf", "type": "function", "stateMutability": "view",
                "inputs": [{"name": "account", "type": "address"}],
                "outputs": [{"name": "", "type": "uint256"}]}]
MIN_SIZE    = 5
BUY_SPEND   = 2.0

w3            = Web3(Web3.HTTPProvider(RPC_URL))
eoa           = w3.eth.account.from_key("0x" + PRIVATE_KEY).address
pusd_contract = w3.eth.contract(address=PUSD, abi=ERC20_ABI)

temp_client = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, signature_type=0)
creds       = temp_client.create_or_derive_api_key()
client      = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, creds=creds, signature_type=0)
client.update_balance_allowance(BalanceAllowanceParams(asset_type=AssetType.COLLATERAL))

event    = requests.get(f"https://gamma-api.polymarket.com/events/{EVENT_ID}").json()
cond_id  = event["markets"][0]["conditionId"]
info     = client.get_clob_market_info(cond_id)
token_id  = next((t["t"] for t in info.get("t", []) if t["o"] == "Yes"), "")
tick_size = str(info.get("mts", "0.01"))
neg_risk  = info.get("nr", False)

book     = client.get_order_book(token_id)
asks     = sorted(book["asks"], key=lambda a: float(a["price"]))
if not asks: print("No asks — can't FOK buy"); exit(1)

best_ask = float(asks[0]["price"])
balance  = pusd_contract.functions.balanceOf(eoa).call() / 1e6
buy_size = max(MIN_SIZE, round(BUY_SPEND / best_ask, 2))
print(f"Best ask: ${best_ask}  pUSD: ${balance:.4f}  size: {buy_size}")

options  = PartialCreateOrderOptions(tick_size=tick_size, neg_risk=neg_risk)
buy_resp = client.create_and_post_order(
    OrderArgsV2(token_id=token_id, price=best_ask, size=buy_size, side="BUY"),
    options, OrderType.FOK,
)
buy_id = buy_resp.get("orderID") if isinstance(buy_resp, dict) else buy_resp.orderID
print(f"Buy: {buy_id}")

# Poll until filled (FOK should fill instantly or fail)
for _ in range(30):
    time.sleep(1)
    ids = [o.get("id", o.get("orderID", "")) for o in (client.get_open_orders() or [])]
    if buy_id not in ids: print("✓ Filled"); break

bids_now = sorted(client.get_order_book(token_id)["bids"], key=lambda b: float(b["price"]), reverse=True)
if not bids_now: print("No bids — skipping sell"); exit(0)

sell_resp = client.create_and_post_order(
    OrderArgsV2(token_id=token_id, price=float(bids_now[0]["price"]), size=buy_size, side="SELL"),
    options, OrderType.FOK,
)
sell_id = sell_resp.get("orderID") if isinstance(sell_resp, dict) else sell_resp.orderID
print(f"Sell: {sell_id}  status: {sell_resp.get('status') if isinstance(sell_resp, dict) else sell_resp.status}")

File: typescript/private_key/6_market_order.ts
// npm run 6:market  — FOK BUY at best ask → poll fill → FOK SELL back
import "dotenv/config";
import { createWalletClient, createPublicClient, http, erc20Abi, formatUnits } from "viem";
import { polygon } from "viem/chains";
import { privateKeyToAccount } from "viem/accounts";
import { ClobClient, Side, OrderType } from "@polymarket/clob-client-v2";

const CLOB_HOST = process.env.CLOB_BASE_URL ?? "https://clob-v2.polymarket.com";
const PUSD      = "0xC011a7E12a19f7B1f670d46F03B03f3342E82DFB" as `0x${string}`;
const EVENT_ID  = "73106";
const MIN_SIZE  = 5;

const account      = privateKeyToAccount(`0x${process.env.PRIVATE_KEY!}` as `0x${string}`);
const walletClient = createWalletClient({ account, chain: polygon, transport: http() });
const publicClient = createPublicClient({ chain: polygon, transport: http() });

const tempClient = new ClobClient({ host: CLOB_HOST, chain: 137, signer: walletClient as never, signatureType: 0 });
const creds      = await (tempClient as any).createOrDeriveApiKey();
const client     = new ClobClient({
  host: CLOB_HOST, chain: 137, signer: walletClient as never, signatureType: 0,
  creds: { key: creds.key, secret: creds.secret, passphrase: creds.passphrase },
});

const event    = await fetch(`https://gamma-api.polymarket.com/events/${EVENT_ID}`).then(r => r.json()) as any;
const condId   = event.markets?.[0]?.conditionId as string;
const info     = await (client as any).getClobMarketInfo(condId);
const tokens   = (info?.t ?? []) as { t: string; o: string }[];
const tokenId  = tokens.find(t => t.o === "Yes")?.t ?? "";
const tickSize = (info?.mts?.toString() ?? "0.01") as "0.1" | "0.01" | "0.001" | "0.0001";
const negRisk  = info?.nr ?? false;

// Best ask + pUSD balance
const book    = await (client as any).getOrderBook(tokenId);
const asks    = [...(book?.asks ?? [])].sort((a: any, b: any) => parseFloat(a.price) - parseFloat(b.price));
const bestAsk = parseFloat(asks[0].price);
const rawBal  = await publicClient.readContract({ address: PUSD, abi: erc20Abi, functionName: "balanceOf", args: [account.address] });
const buySize = Math.max(MIN_SIZE, Math.floor((2 / bestAsk) * 100) / 100);

// FOK market BUY
const buyResp = await client.createAndPostOrder(
  { tokenID: tokenId, price: bestAsk, size: buySize, side: Side.BUY },
  { tickSize, negRisk }, OrderType.FOK,
);
console.log(`Buy: ${buyResp.orderID}  status: ${buyResp.status}`);

// Poll until filled
for (let i = 0; i < 30; i++) {
  await new Promise(r => setTimeout(r, 1000));
  const open: any[] = await (client as any).getOpenOrders() ?? [];
  if (!open.find((o: any) => o.id === buyResp.orderID)) { console.log("✓ Filled"); break; }
}

// FOK market SELL back
const bids    = [...(book?.bids ?? [])].sort((a: any, b: any) => parseFloat(b.price) - parseFloat(a.price));
const bestBid = parseFloat(bids[0].price);
const sellResp = await client.createAndPostOrder(
  { tokenID: tokenId, price: bestBid, size: buySize, side: Side.SELL },
  { tickSize, negRisk }, OrderType.FOK,
);
console.log(`Sell: ${sellResp.orderID}  status: ${sellResp.status}`);
